libraries/AP_RangeFinder/AP_RangeFinder_GP2Y0E03.cpp

```diff
@@ -0,0 +1,161 @@
+        AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+    : AP_RangeFinder_Backend(_state, _params)
+    , _dev(std::move(dev))
+{
+}
+
+AP_RangeFinder_Backend *AP_RangeFinder_GP2Y0E03::detect(
+        RangeFinder::RangeFinder_State &_state, AP_RangeFinder_Params &_params,
+        AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+{
+    if (!dev) {
+        return nullptr;
+    }
+
+    AP_RangeFinder_GP2Y0E03 *sensor
+        = new AP_RangeFinder_GP2Y0E03(_state, _params, std::move(dev));
+
+    if (!sensor ) {
+        delete sensor;
+        return nullptr;
+    }
+    
+    if (!sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+bool AP_RangeFinder_GP2Y0E03::init()
+{
+    uint8_t val;
+    bool ret;
+
+    const uint8_t CMD_SYSTEM_RESET = 0xee;    
+    const uint8_t CMD_SOFTWARE_RESET = 0x06;
+    const uint8_t CMD_PULSE_WIDTH = 0x13;
+
+    _dev->get_semaphore()->take_blocking();
+    _dev->set_retries(10);
+
+    /*
+     * Reseting the sensor
+     */
+    ret = _dev->write_register(CMD_SYSTEM_RESET, CMD_SOFTWARE_RESET);
+    if (!ret) {
+        goto fail;
+    }
+
+    hal.scheduler->delay(100);
+
+    //Reading the attribute Width to check if the sensor is alive
+    _dev->set_retries(10);    
+    ret = _dev->read_registers(CMD_PULSE_WIDTH, &val,sizeof(val));
+    if (!ret) {
+        goto fail;
+    }
+    
+    if (val != 7) {
+        GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "GP2Y0E03: ERROR! %d", val);
+        goto fail;
+    }
+
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "GP2Y0E03: Pulse Width: %d",val);
+    
+    _dev->get_semaphore()->give();
+
+    hal.scheduler->delay(100);
+
+    _dev->register_periodic_callback(20000,
+                                     FUNCTOR_BIND_MEMBER(&AP_RangeFinder_GP2Y0E03::timer, void));
+
+    return true;
+
+fail:
+    _dev->get_semaphore()->give();
+    return false;
+}
+
+void AP_RangeFinder_GP2Y0E03::update()
+{
+    WITH_SEMAPHORE(_sem);
+
+    if (accum.count > 0) {
+        state.distance_cm = accum.sum / accum.count;
+        state.last_reading_ms = AP_HAL::millis();
+        accum.sum = 0;
+        accum.count = 0;
+        update_status();
+    } else if (AP_HAL::millis() - state.last_reading_ms > 200) {
+        set_status(RangeFinder::Status::NoData);
+    }
+}
+
+void AP_RangeFinder_GP2Y0E03::timer()
+{
+    const uint8_t CMD_READ_DISTANCE_1 = 0x5e;
+    const uint8_t CMD_READ_DISTANCE_2 = 0x5f;
+    const uint8_t CMD_READ_SHIFT = 0x35;
+    
+    uint8_t shift, high, low;
+    uint16_t data;
+    
+    bool ret;
+    
+    ret = _dev->read_registers(CMD_READ_SHIFT, &shift, sizeof(shift));
+    if(!ret) {
+        return;
+    }
+    
+    ret = _dev->read_registers(CMD_READ_DISTANCE_1, &high, sizeof(high));
+    if (!ret) {
+        return;
+    }
+    
+    ret = _dev->read_registers(CMD_READ_DISTANCE_2, &low, sizeof(low));
+    if (!ret) {
+        return;
+    }
+    
+    data = (high << 4) | low;
+    data = (data / 16) >> shift;
```

Comment:
```suggestion
    const uint16_t data = (((high << 4) | low) / 16) >> shift;
```

I haven't vetted this maths (assuming you've tested it and it gives you the right data.  But given division by a power of 2 is the same as a right-shift-by-n-bits this doesn't look optimal, at the very least (division is expensive!)

Reply:
The distance measurement of this device is 4 to 50 centimeters.
I think I need a valid value determination.

https://global.sharp/products/device/lineup/data/pdf/datasheet/gp2y0e02_03_appl_e.pdf